File: grimdb/sdk/java/src/main/java/com/grimlocker/sdk/GrimlockerException.java
package com.grimlocker.sdk;

/** Thrown when the Grimlocker daemon returns an error or the connection fails. */
public class GrimlockerException extends RuntimeException {

    private final int errorCode;

    public GrimlockerException(String message) {
        super(message);
        this.errorCode = 0;
    }

    public GrimlockerException(int errorCode, String errorName, String errorMsg) {
        super(errorName + " (" + errorCode + "): " + errorMsg);
        this.errorCode = errorCode;
    }

    public GrimlockerException(String message, Throwable cause) {
        super(message, cause);
        this.errorCode = 0;
    }

    /** The GQL numeric error code, or 0 if this is a connection/protocol error. */
    public int getErrorCode() {
        return errorCode;
    }
}

File: grimdb/sdk/kotlin/src/main/kotlin/com/grimlocker/sdk/GrimlockerException.kt
package com.grimlocker.sdk

class GrimlockerException(
    val errorCode: Int,
    message: String
) : RuntimeException(message) {

    constructor(message: String) : this(0, message)

    override fun toString(): String = "GrimlockerException(code=$errorCode, message='$message')"

    companion object {
        fun nameOf(code: Int): String = when (code) {
            -1   -> "BUS_ERROR"
            -2   -> "STORAGE_ERROR"
            -3   -> "NOT_FOUND"
            -10  -> "ENTRY_NOT_FOUND"
            -20  -> "CATEGORY_ERROR"
            -30  -> "CREATE_FAILED"
            -31  -> "UPDATE_FAILED"
            -32  -> "DELETE_FAILED"
            -100 -> "PROTOCOL_ERROR"
            -101 -> "AUTH_REQUIRED"
            -102 -> "PERMISSION_DENIED"
            -103 -> "INVALID_REQUEST"
            -104 -> "TIMEOUT"
            else -> "UNKNOWN"
        }
    }
}
